strongbox-security/strongbox-security-api/src/main/java/org/carlspring/strongbox/configuration/Base64PasswordEncoderDelegate.java

```diff
@@ -0,0 +1,207 @@
+package org.carlspring.strongbox.configuration;
+
+import java.util.Base64;
+
+import org.apache.commons.lang3.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.crypto.codec.Hex;
+import org.springframework.security.crypto.codec.Utf8;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+/**
+ * This class encodes given raw String or Base64 encoded string.
+ * If given string is Base64 Encoded it will attempt to decode the password and finally delegate it to the password encoder.
+ * This handles all possible cases
+ * <p>
+ * {ALG}md5/sha1/bcrypt(mypassword)
+ * {ALG}base64.encode(md5/sha1/bcrypt(mypassword))
+ * base64.encode({ALG}md5/sha1/bcrypt(mypassword))
+ *
+ * @author mbharti
+ * @date 07/10/20
+ */
+@Component
+public class Base64PasswordEncoderDelegate
+        implements PasswordEncoder
+{
+
+    private static final String PREFIX = "{";
+
+    private static final String SUFFIX = "}";
+
+    private final PasswordEncoder passwordEncoder;
+
+    private static final Logger logger = LoggerFactory.getLogger(Base64PasswordEncoderDelegate.class);
+
+
+    /**
+     * Constructs PasswordEncoder with given PasswordEncoder
+     *
+     * @param passwordEncoder
+     */
+    public Base64PasswordEncoderDelegate(PasswordEncoder passwordEncoder)
+    {
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    /**
+     * This Encodes the raw password. In case given rawCharSequence is Base64 encoded,
+     * then it will decode and then encode the decoded string.
+     *
+     * @param rawCharSequence Raw String or Base64 encoded String
+     * @return Encoded String
+     */
+    @Override
+    public String encode(CharSequence rawCharSequence)
+    {
+        return passwordEncoder.encode(rawCharSequence);
```

Comment:
maybe we should do Base 64 encode here as well?

Reply:
if we do Base64Enocding here, will it get applied to all passwords encoded by the underlying framework? In that case, we will be adding extra one more level of encoding decoding?
Also while decoding we just decode with the given raw password and encoded password
if that fails then all other possible checks start.

- Currently, if a password generated by this framework, it will pass at first go itself without doing any encoding decoding. In case it fails it will try the other two possible scenarios 


Please share your thoughts on this, we can change it accordingly

Reply:
thanks for clarifying!
I seems missed that this is just common `PasswordEncoder` that we use for all purposes. Considering that it would be good to apply this base 64 related logic only for LDAP and this seems not possible with `PasswordEncoder` in our case. So I'm just noticed that we could customize `LdapUserDetailsMapper.mapPassword` for this purpose. 
@madhukarbharti,  could you please try to proceed this way? The suggestion is to have customized `LdapUserDetailsMapper` implementation with optional base 64 decode logic. This could be configurable in `/strongbox-ldap-authentication-provider/src/main/resources/org/carlspring/strongbox/authentication/external/ldap/strongbox-authentication-providers.xml` with `strongbox.authentication.ldap.base64EncodedPassword=true|false` property.